providers/implementations/kdfs/argon2.c

```diff
@@ -0,0 +1,1531 @@
+                   const void *in, size_t inlen, const void *key, size_t keylen)
+{
+    if (out == NULL || outlen == 0)
+        return 0;
+
+    if (key == NULL || keylen == 0)
+        return blake2b_md(md, out, outlen, in, inlen);
+
+    return blake2b_mac(mac, out, outlen, in, inlen, key, keylen);
+}
+
+static int blake2b_long(EVP_MD *md, EVP_MAC *mac, unsigned char *out,
+                        size_t outlen, const void *in, size_t inlen)
+{
+    int ret = 0;
+    EVP_MD_CTX *ctx = NULL;
+    uint32_t outlen_curr;
+    uint8_t outbuf[BLAKE2B_OUTBYTES];
+    uint8_t inbuf[BLAKE2B_OUTBYTES];
+    uint8_t outlen_bytes[sizeof(uint32_t)] = {0};
+    OSSL_PARAM par[2];
+    size_t outlen_md;
+
+    if (out == NULL || outlen == 0)
+        return 0;
+
+    /* Ensure little-endian byte order */
+    store32(outlen_bytes, (uint32_t)outlen);
+
+    if ((ctx = EVP_MD_CTX_create()) == NULL)
+        return 0;
+
+    outlen_md = (outlen <= BLAKE2B_OUTBYTES) ? outlen : BLAKE2B_OUTBYTES;
+    par[0] = OSSL_PARAM_construct_size_t(OSSL_DIGEST_PARAM_XOFLEN, &outlen_md);
+    par[1] = OSSL_PARAM_construct_end();
+
+    ret = EVP_DigestInit_ex2(ctx, md, par) == 1
+        && EVP_DigestUpdate(ctx, outlen_bytes, sizeof(outlen_bytes)) == 1
+        && EVP_DigestUpdate(ctx, in, inlen) == 1
+        && EVP_DigestFinalXOF(ctx, (outlen > BLAKE2B_OUTBYTES) ? outbuf : out,
+                outlen_md) == 1;
+
+    if (ret == 0)
+        goto fail;
+
+    if (outlen > BLAKE2B_OUTBYTES) {
+        memcpy(out, outbuf, BLAKE2B_OUTBYTES / 2);
+        out += BLAKE2B_OUTBYTES / 2;
+        outlen_curr = (uint32_t) outlen - BLAKE2B_OUTBYTES / 2;
+
+        while (outlen_curr > BLAKE2B_OUTBYTES) {
+            memcpy(inbuf, outbuf, BLAKE2B_OUTBYTES);
+            if (blake2b(md, mac, outbuf, BLAKE2B_OUTBYTES, inbuf,
+                        BLAKE2B_OUTBYTES, NULL, 0) != 1)
+                goto fail;
+            memcpy(out, outbuf, BLAKE2B_OUTBYTES / 2);
+            out += BLAKE2B_OUTBYTES / 2;
+            outlen_curr -= BLAKE2B_OUTBYTES / 2;
+        }
+
+        memcpy(inbuf, outbuf, BLAKE2B_OUTBYTES);
+        if (blake2b(md, mac, outbuf, outlen_curr, inbuf, BLAKE2B_OUTBYTES,
+                    NULL, 0) != 1)
+            goto fail;
+        memcpy(out, outbuf, outlen_curr);
+    }
+    ret = 1;
+
+fail:
+    EVP_MD_CTX_free(ctx);
+    return ret;
+}
+
+static void kdf_argon2_init(KDF_ARGON2 *c, ARGON2_TYPE type)
+{
+    OSSL_LIB_CTX *libctx;
+
+    libctx = c->libctx;
+    memset(c, 0, sizeof(*c));
+    c->libctx = libctx;
+    c->outlen = ARGON2_DEFAULT_OUTLEN;
+    c->t_cost = ARGON2_DEFAULT_T_COST;
+    c->m_cost = ARGON2_DEFAULT_M_COST;
+    c->lanes = ARGON2_DEFAULT_LANES;
+    c->threads = ARGON2_DEFAULT_THREADS;
+    c->version = ARGON2_DEFAULT_VERSION;
+    c->early_clean = 0;
+    c->type = type;
+    c->mac = EVP_MAC_fetch(c->libctx, "blake2bmac", NULL);
+    c->md = EVP_MD_fetch(c->libctx, "blake2b512", NULL);
```

Comment:
These fetches should be checked for NULL return. Also there should be a way to set the propquery as it is done for other KDFs when fetching the used digests or ciphers. I.e. the propquery is being passed in as a parameter OSSL_ALG_PARAM_PROPERTIES see ossl_prov_digest_load_from_params(). It's different in this case that the digest itself is not changeable but the propquery stil should be IMO.